Instance ID Picking

danieldresser-ie opened the pull request:
This initial implementation of image picking for instance ids should be ready for review.

On the renderer side, it's a bit weird because we want this to theoretically work across all renderers, but only Arnold actually supports encapsulated instancers.

I'm also not quite fully sure how we communicate to users that this new mode is only used for instance ids for encapsulated instancers, and it won't show any information on non-encapsulated renders ( this behaviour seems correct, but might not immediately be intuitive ).

Other than that, the main ugly bit is that we need to handle 0 ids here - I haven't squashed the two fix commits that implement this yet, because I'm not sure yet we've got the right approach. I've tried to make the offset-by-one feel like part of the bitcast translation from a float ... but we don't want this toggle to impact ImageGadget, so ImageGadget does the bitcast without an offset, and we instead offset the ids being passed in ... which works, but feels inconsistent. From this point of view, it would be better if we just passed in floats to match with, and ImageGadget didn't know that ids were integers ( it just go a list of floats to match with ) - but I think I moved away from that because doing floating point comparisons on the GPU left us vulnerable to floating point comparison weirdness.

The idea of always including an offset-by-one as part of our int->float conversion is kind of attractive to me still? I think your argument against it was that you didn't like preserving zeros for the main id pass because RenderManifest::acquireID never returns 0, so it's not necessary to preserve zeros for the main id pass?

johnhaddon reviewed (commented):
Thanks Daniel - works good. Comments inline as usual, and replies to your thoughts below...

> I'm also not quite fully sure how we communicate to users that this new mode is only used for instance ids for encapsulated instancers, and it won't show any information on non-encapsulated renders ( this behaviour seems correct, but might not immediately be intuitive ).

Yeah, it is a tricky one. I don't think the "within encapsulated instancer" part of the status will help at all at Cinesite, where the vast majority of instancers are USDGeomPointInstancer. I'm tempted to leave that bit out even. Things will make a lot more sense when we move to passing around first-class point instancers and rebill Instancer more accurately.

> I've tried to make the offset-by-one feel like part of the bitcast translation from a float ... 

Conceptually I think those are reasonably separate things (data type and background value), so I'm not sure that making them feel like one thing is all that beneficial.

> we don't want this toggle to impact ImageGadget

I think the thing I _really_ wanted to avoid was ImageGadget having any concept of ID vs InstanceID, or specific meanings for channel names. I could probably live with an `ImageGadget::setIDOffset()` mode though, if it helps make anything clearer. I did make one suggestion that I think will make everything feel a bit cleaner anyway though - maybe see how you feel after that?

Cheers...
John

danieldresser-ie commented on src/GafferSceneUI/ImageSelectionTool.cpp:
```diff
@@ -704,14 +789,30 @@ IECore::RunTimeTypedPtr ImageSelectionTool::dragBegin( GafferUI::Gadget *gadget,
+			}
+			else
+			{
+				IECore::Int64VectorDataPtr dragData = new IECore::Int64VectorData();
+				std::vector< int64_t > &dragWritable = dragData->writable();
+				for( const uint32_t &i : m_selectedIDs )
+				{
+					dragWritable.push_back( i );
+				}
```
Int64Data is for compatibility with the "idList" plug on PrimitiveVariableTweaks, which is the most obvious place to  use this data ( if you want to deactivate some instances ). Ideally, that plug would probably support drags of any integer vector data type, and I wouldn't need to do this cast, but it seemed like the drag handling for this was handled kind of centrally, so rather than making the drag acceptance more flexible, I just used the compatible type. Not exactly sure what you mean by "insert a range" -  are you thinking something like `dragWritable.insert( m_selectedIDs.begin(), m_selectedIDs.end() );`? That just generates a bunch of errors about the iterator types not matching.

I could certainly just add a `reserve()` if it the performance you're concerned about.

johnhaddon replied:
> Ideally, that plug would probably support drags of any integer vector data type, and I wouldn't need to do this cast

Yeah, it would be much better for the destination to be flexible in what it accepts rather than having the source do a conversion based on assumptions about what the destination is.

Looks like a special case in `VectorDataWidget.__dragEnter` would be sufficient to get this working without doing the conversion in ImageSelectionTool - could you give that a go?

> are you thinking something like dragWritable.insert( m_selectedIDs.begin(), m_selectedIDs.end() )

Yep, that's what I was thinking - to reduce the verbosity as much as anything else. If we can omit the conversion this will just be a one-liner anyway...

danieldresser-ie commented on src/GafferSceneUI/ImageSelectionTool.cpp:
```diff
@@ -212,11 +225,13 @@ ImageSelectionTool::ImageSelectionTool( View *view, const std::string &name )
 
 	view->viewportGadget()->preRenderSignal().connect( boost::bind( &ImageSelectionTool::preRender, this ) );
 	plugDirtiedSignal().connect( boost::bind( &ImageSelectionTool::plugDirtied, this, ::_1 ) );
+	plugSetSignal().connect( boost::bind( &ImageSelectionTool::plugSet, this, ::_1 ) );
```
Not sure what was going wrong when I tested this, but I've now moved everything back to plugDirtied instead of plugSet and it all seems to be working fine.

danieldresser-ie replied:
Addressed in 2f72b02ebe7c2ff5e8e03418e699af31634ea40f

johnhaddon commented on python/GafferSceneUI/ImageSelectionToolUI.py:
```diff
@@ -110,12 +130,15 @@ def __init__( self, tool, **kw ) :
+			self._qtWidget().setObjectName( "gafferImageSelectionStatus" )
 
-		self.__tool.statusChangedSignal().connect( Gaffer.WeakMethod( self.__update, fallbackResult = None ) )
+			self.__tool.statusChangedSignal().connect( Gaffer.WeakMethod( self.__update, fallbackResult = None ) )
```
Repro steps :

1. Batch render an instanceID output, making sure to omit a manifest.
2. Load it in with an ImageReader, and view it.
3. Activate the ImageSelectionTool. The error will show, but that's OK, because we're in "Standard" mode and a manifest is needed for that.
4. Switch mode to "Instance". The error remains.

danieldresser-ie replied:
Both addressed in 70514c6187e14a0995dc397c17fbfab9e4bcc942

danieldresser-ie commented on src/GafferScene/Catalogue.cpp:
```diff
@@ -634,17 +634,22 @@ class Catalogue::InternalImage : public ImageNode
+
+					Spreadsheet::RowPlug *instanceRow = spreadsheet->rowsPlug()->addRow();
+					instanceRow->namePlug()->setValue( "instanceID" );
+					instanceRow->cellsPlug()->getChild<Spreadsheet::CellPlug>(0)->valuePlug<StringPlug>()->setValue( "float" );
```
Fixed in 1498975dd2c3beba4d02dd49ab052f97c17da5ad

danieldresser-ie commented on src/IECoreDelight/Renderer.cpp:
```diff
@@ -368,12 +368,20 @@ class DelightOutput : public IECore::RefCounted
+				// \todo : It feels messy to use upper case ID here, but "Id" for the actual channel
+				// name in the image. But it would perhaps be even weirder to correct this for
+				// cortexInstanceId and not cortexId and fixing cortexId could break compability with
+				// existing image files ... so maybe we're stuck with this?
```
Addressed in 67fdcf5ae46c0a57ac3c31ef780505457a5ddd62 and 70a78c860e27c9442dda85e08af642e10a996afa

danieldresser-ie commented on src/IECoreDelight/Renderer.cpp:
```diff
@@ -1187,6 +1226,7 @@ class DelightObject: public IECoreScenePreview::Renderer::ObjectInterface
 
 		DelightHandleSharedPtr m_instance;
 		DelightHandle m_idAttributesHandle;
+		DelightHandle m_instanceIDAttributesHandle;
```
Addressed in a2986dcf991ce2b554c174da2ffc2237e1244fc6

danieldresser-ie commented on src/IECoreArnold/Renderer.cpp:
```diff
@@ -2594,8 +2594,9 @@ class ArnoldObjectBase : public IECoreScenePreview::Renderer::ObjectInterface
+					uint32_t instanceIDPlusOne = instanceID + 1;
 					float bitcastID;
-					memcpy( &bitcastID, &instanceID, 4 );
+					memcpy( &bitcastID, &instanceIDPlusOne, 4 );
```
Addressed in 8c5674bf1e85cf74c624ff193641ed4e9e3047c0

danieldresser-ie commented on python/GafferSceneUI/ImageSelectionToolUI.py:
```diff
@@ -82,26 +81,47 @@
+			"description",
+			"""
+			Determines the scene location that is ultimately selected or deselected,
+			which may differ from what is originally selected.
```
Addressed in 49ac4f845a9a5a95e7b65e435064340b1fa8f278

danieldresser-ie commented on python/GafferSceneUI/ImageSelectionToolUI.py:
```diff
@@ -82,26 +81,47 @@
+			""",
+
+			"plugValueWidget:type", "GafferUI.PresetsPlugValueWidget",
+			"preset:Standard", "",
```
Addressed in 49ac4f845a9a5a95e7b65e435064340b1fa8f278

danieldresser-ie commented on python/GafferSceneUI/ImageSelectionToolUI.py:
```diff
@@ -130,18 +153,18 @@ def __update( self, *unused ) :
 
 		status = self.__tool.status()
-		self.__frame.setVisible( bool(status) )
+		#self.__frame.setVisible( bool(status) )
```
Removed debug code in https://github.com/GafferHQ/gaffer/commit/49ac4f845a9a5a95e7b65e435064340b1fa8f278

danieldresser-ie commented on src/GafferSceneUI/ImageSelectionTool.cpp:
```diff
@@ -260,15 +277,29 @@ std::string ImageSelectionTool::status() const
-		return "info:Hover object to show path";
+		if( instanceSelection )
+		{
+			return "info:Hover instance within encapsulated instancer to id";
```
Fixed in d6e2d72349456365431130723075e4ffaa33de0d

danieldresser-ie commented on src/GafferSceneUI/ImageSelectionTool.cpp:
```diff
@@ -578,7 +597,7 @@ std::unordered_set<uint32_t> ImageSelectionTool::rectIDs( const Imath::Box2i &re
 				if( id != 0 )
 				{
-					result.insert( id );
+					result.insert( id - ( instance ? 1 : 0 ) );
```
We don't actually need an `optional`, because we know that it's a background value whenever the raw int is equal to 0, but I was able to get rid of the duplicated init code in 476db0e015d52094f7927842bb27cc366a58d47b

danieldresser-ie commented:
Added two more commits based on our discussion this morning. Fixed the path issue in the Delight test on Windows ( the Linux tests passed this time, so guess it was random ). And did some UI hacking to try and clean things up. I did get the label height reduced ... but things still aren't lining up vertically exactly right. I tried a few different hacks to try and understand things, but without much success. ( I can get the alignment better if I change the AlignTop on line 97 of PlugWidget to an AlignVCenter, but it still doesn't exactly match other things, and there's no exposed way to set that alignment ... I'm guessing that somehow I'm supposed to have the PlugWidget inside a vertically limited container so it doesn't matter what the vertical alignment is? )

johnhaddon commented:
> things still aren't lining up vertically exactly right

This was doing my nut in so I've pushed https://github.com/GafferHQ/gaffer/pull/6503/commits/44ad31716bb21397a89a68c579e724737e1608de which contains the minimal change necessary to sort this out.

I've also pushed https://github.com/GafferHQ/gaffer/pull/6503/commits/db27f5bea9e23328fbbe3ae88c583daec13a7eb8 which simplifies things further while keeping the same layout.

Could you get everything squashed ready for merging please?

danieldresser-ie commented:
Squashed, added some basic Changelog entries.
